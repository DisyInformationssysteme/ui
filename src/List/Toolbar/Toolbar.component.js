import React from 'react';
import Navbar from 'react-bootstrap/lib/Navbar';
import SelectDisplayMode from './SelectDisplayMode';
import SelectSortBy from './SelectSortBy';
import Pagination from './Pagination';
import Filter from './Filter';
import ActionBar from '../../ActionBar';

import theme from './Toolbar.scss';

export function getSubProps(props, component) {
	const subProps = {};
	Object.keys(component.propTypes)
		.filter(key => props[key] !== undefined)
		.forEach((key) => {
			subProps[key] = props[key];
		});
	return subProps;
}

function adaptActionsIds(actions, parentId) {
	return actions &&
		actions.map((action) => {
			if (action.id) {
				return {
					...action,
					id: `${parentId}-${action.id}`,
				};
			}
			return action;
		});
}

function hasProps(props) {
	return Object.keys(props).find(key => key !== 'id');
}

/**
 * @param {object} props react props
 * @example
 <Toolbar name="Hello world"></Toolbar>
 */
function Toolbar(props) {
	const displayProps = getSubProps(props, SelectDisplayMode);
	const sortProps = getSubProps(props, SelectSortBy);
	const filterProps = getSubProps(props, Filter);
	const paginationProps = getSubProps(props, Pagination);

	let actions = props.actions;
	if (props.id && actions) {
		actions = {
			left: adaptActionsIds(actions.left, props.id),
			right: adaptActionsIds(actions.right, props.id),
		};
	}

	const hasDisplayMode = hasProps(displayProps);
	const hasSortProps = hasProps(sortProps);
	const hasFilterProps = hasProps(filterProps);
	const hasPaginationProps = hasProps(paginationProps);

	return (
		<div>
			{actions && (<ActionBar actions={actions} />)}
			<Navbar componentClass="div" className={theme['tc-list-toolbar']} role="toolbar" fluid>
				{hasDisplayMode && (<SelectDisplayMode {...displayProps} />)}
				{hasSortProps && (<SelectSortBy {...sortProps} />)}
				{hasFilterProps && (<Filter {...filterProps} />)}
				{hasPaginationProps && (<Pagination {...paginationProps} />)}
			</Navbar>
		</div>
	);
}

Toolbar.propTypes = {
	id: React.PropTypes.string,
	...Filter.propTypes,
	...SelectDisplayMode.propTypes,
	...SelectSortBy.propTypes,
};

export default Toolbar;
