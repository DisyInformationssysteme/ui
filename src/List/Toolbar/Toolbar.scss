// Toolbar styles
$tc-list-toolbar-display-color: $silver !default;
$tc-list-toolbar-background-color: $concrete !default;


.tc-list-toolbar {
	background: $tc-list-toolbar-background-color;

	:global(.navbar-text) {
		color: $tc-list-toolbar-display-color;
		font-weight: 400;
		margin-right: 0;
	}
}
