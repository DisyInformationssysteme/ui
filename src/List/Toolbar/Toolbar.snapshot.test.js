import React from 'react';
import renderer from 'react-test-renderer';

import Toolbar, { getSubProps } from './Toolbar.component';

jest.mock('react-dom');

const props = {
	onSelectDisplayMode: jest.fn(),
	onSelectSortBy: jest.fn(),
	onFilter: jest.fn(),
	sortBy: [
		{ id: 'id', name: 'Name', selected: true },
		{ id: 'name', name: 'Name' },
	],
	sortDesc: true,
	itemsLength: 17,
	onChangePagination: jest.fn(),
};

describe('Toolbar', () => {
	it('should render', () => {
		const wrapper = renderer.create(
			<Toolbar {...props} />
		).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
	it('should getSubProps', () => {
		const propTypes = {
			onSelectDisplayMode: 'func',
			sortBy: 'bool',
		};
		const subProps = getSubProps(props, { propTypes });
		expect(subProps.onSelectDisplayMode).toBe(props.onSelectDisplayMode);
		expect(subProps.sortBy).toBe(props.sortBy);
		expect(Object.keys(subProps).length).toBe(2);
	});
	it('should render listActions', () => {
		const tprops = Object.assign({}, props);
		tprops.listActions = [
			{
				label: 'Delete selection',
				icon: 'fa fa-trash-o',
				onClick: jest.fn(),
			},
		];
		const wrapper = renderer.create(
			<Toolbar {...tprops} />
		).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
	it('should render add btn', () => {
		const tprops = Object.assign({}, props);
		tprops.onClickAdd = jest.fn();
		const wrapper = renderer.create(
			<Toolbar {...tprops} />
		).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
});
