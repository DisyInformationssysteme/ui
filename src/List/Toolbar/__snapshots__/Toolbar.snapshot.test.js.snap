exports[`Toolbar should render actions toolbar 1`] = `
<div>
  <nav
    className="tc-actionbar-container nav">
    <div
      className="navbar-left">
      <button
        className="btn btn-primary"
        disabled={false}
        id="my-toolbar-add"
        onClick={[Function]}
        role={null}
        type="button">
        <svg
          aria-hidden="true"
          className="tc-svg-icon"
          title={null}>
          <use
            xlinkHref="#talend-plus" />
        </svg>
        <span>
          Add
        </span>
      </button>
      <button
        className="btn btn-default"
        disabled={false}
        id="my-toolbar-delete"
        onClick={[Function]}
        role={null}
        type="button">
        <svg
          aria-hidden="true"
          className="tc-svg-icon"
          title={null}>
          <use
            xlinkHref="#talend-trash" />
        </svg>
        <span>
          Delete selection
        </span>
      </button>
    </div>
  </nav>
  <div
    className="navbar navbar-default"
    role="toolbar">
    <div
      className="container-fluid" />
  </div>
</div>
`;

exports[`Toolbar should render display mode selector 1`] = `
<div>
  <div
    className="navbar navbar-default"
    role="toolbar">
    <div
      className="container-fluid">
      <ul
        className="nav navbar-nav"
        role={null}>
        <label
          active={undefined}
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-text"
          htmlFor="my-toolbar-display-mode"
          onSelect={null}>
          Display:
        </label>
        <li
          className="dropdown"
          style={undefined}>
          <a
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle"
            href=""
            id="my-toolbar-display-mode"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button">
            <i
              aria-hidden="true"
              className="fa fa-bars"
              title={null} />
             
            <span
              className="caret" />
          </a>
          <ul
            aria-labelledby="my-toolbar-display-mode"
            className="dropdown-menu"
            role="menu">
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="my-toolbar-display-mode-table"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                <i
                  aria-hidden="true"
                  className="fa fa-bars"
                  title={null} />
                 
                Table
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="my-toolbar-display-mode-large"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                <i
                  aria-hidden="true"
                  className="fa fa-list-alt"
                  title={null} />
                 
                Expanded
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="my-toolbar-display-mode-tile"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                <i
                  aria-hidden="true"
                  className="fa fa-th-large"
                  title={null} />
                 
                Tile
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Toolbar should render empty toolbar 1`] = `
<div>
  <div
    className="navbar navbar-default"
    role="toolbar">
    <div
      className="container-fluid" />
  </div>
</div>
`;

exports[`Toolbar should render filter form 1`] = `
<div>
  <div
    className="navbar navbar-default"
    role="toolbar">
    <div
      className="container-fluid">
      <form
        className="navbar-form navbar-right"
        onSubmit={[Function]}
        role="search">
        <div
          className="form-group">
          <input
            aria-label="Filter"
            className="form-control"
            id="my-toolbar-filter-input"
            onChange={[Function]}
            placeholder="Filter"
            type="search" />
          <button
            className="btn btn-link"
            disabled={false}
            id="my-toolbar-filter-submit"
            title="Filter"
            type="submit">
            <svg
              aria-hidden="true"
              className="tc-svg-icon"
              title={null}>
              <use
                xlinkHref="#talend-search" />
            </svg>
            <span
              className="sr-only">
              Filter
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Toolbar should render pagination 1`] = `
<div>
  <div
    className="navbar navbar-default"
    role="toolbar">
    <div
      className="container-fluid">
      <ul
        className="nav navbar-nav"
        role={null}>
        <label
          active={undefined}
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-text"
          htmlFor="my-toolbar-pagination-size"
          onSelect={null}>
          Show:
        </label>
        <li
          className="dropdown"
          style={undefined}>
          <a
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle"
            href=""
            id="my-toolbar-pagination-size"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button">
            5
             
            <span
              className="caret" />
          </a>
          <ul
            aria-labelledby="my-toolbar-pagination-size"
            className="dropdown-menu"
            role="menu">
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                5
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                10
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                20
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                50
              </a>
            </li>
          </ul>
        </li>
        <div
          active={undefined}
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-btn btn-group"
          onSelect={null}>
          <button
            className="btn btn-link"
            disabled={false}
            id="my-toolbar-nav-to-first"
            onClick={[Function]}
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-backward"
              title={null} />
          </button>
          <button
            className="btn btn-link tc-pagination-ctrl-prev btn btn-link tc-pagination-ctrl-prev-default"
            disabled={false}
            id="my-toolbar-nav-to-prev"
            onClick={[Function]}
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-play"
              title={null} />
          </button>
          <span
            className="btn btn-link">
            6
            /
            2
          </span>
          <button
            className="btn btn-link"
            disabled={false}
            id="my-toolbar-nav-to-next"
            onClick={[Function]}
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-play"
              title={null} />
          </button>
          <button
            className="btn btn-link"
            disabled={false}
            id="my-toolbar-nav-to-last"
            onClick={[Function]}
            type="button">
            <i
              aria-hidden="true"
              className="fa fa-forward"
              title={null} />
          </button>
        </div>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Toolbar should render sort selector 1`] = `
<div>
  <div
    className="navbar navbar-default"
    role="toolbar">
    <div
      className="container-fluid">
      <ul
        className="nav navbar-nav"
        role={null}>
        <label
          active={undefined}
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-text"
          htmlFor="my-toolbar-sort-by"
          onSelect={null}>
          Sort by:
        </label>
        <li
          className="dropdown"
          style={undefined}>
          <a
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle"
            href=""
            id="my-toolbar-sort-by"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button">
            Id
             
            <span
              className="caret" />
          </a>
          <ul
            aria-labelledby="my-toolbar-sort-by"
            className="dropdown-menu"
            role="menu">
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="my-toolbar-sort-by-item-Id"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Id
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                id="my-toolbar-sort-by-item-Name"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Name
              </a>
            </li>
          </ul>
        </li>
        <button
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-btn btn btn btn-link"
          disabled={false}
          id="my-toolbar-sort-order"
          onClick={[Function]}
          onSelect={null}
          role="button"
          type="button">
          DESCENDING
        </button>
      </ul>
    </div>
  </div>
</div>
`;
